library/src/main/java/com/datatorrent/lib/join/AbstractManagedStateInnerJoinOperator.java

```diff
@@ -0,0 +1,354 @@
+      return;
+    }
+    Spillable.SpillableByteArrayListMultimap<K, T> valuestore = isStream1Data ? stream2Data : stream1Data;
+    Future<List> future = ((ManagedTimeStateMultiValue)valuestore).getAsync(key);
+    if (future.isDone()) {
+      try {
+        joinStream(tuple,isStream1Data, future.get());
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      } catch (ExecutionException e) {
+        throw new RuntimeException(e);
+      }
+    } else {
+      waitingEvents.put(new JoinEvent<>(key,tuple,isStream1Data),future);
+    }
+  }
+
+  @Override
+  public void handleIdleTime()
+  {
+    if (waitingEvents.size() > 0) {
+      processWaitEvents(false);
+    }
+  }
+
+  @Override
+  public void beforeCheckpoint(long l)
+  {
+    stream1Store.beforeCheckpoint(l);
+    stream2Store.beforeCheckpoint(l);
+  }
+
+  @Override
+  public void checkpointed(long l)
+  {
+    stream1Store.checkpointed(l);
+    stream2Store.checkpointed(l);
+  }
+
+  @Override
+  public void committed(long l)
+  {
+    stream1Store.committed(l);
+    stream2Store.committed(l);
+  }
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    super.setup(context);
+    ((FileAccessFSImpl)stream1Store.getFileAccess()).setBasePath(context.getValue(DAG.APPLICATION_PATH) + Path.SEPARATOR + stateDir + Path.SEPARATOR + String.valueOf(context.getId()) + Path.SEPARATOR + stream1State);
+    ((FileAccessFSImpl)stream2Store.getFileAccess()).setBasePath(context.getValue(DAG.APPLICATION_PATH) + Path.SEPARATOR + stateDir + Path.SEPARATOR + String.valueOf(context.getId()) + Path.SEPARATOR + stream2State);
+    stream1Store.getCheckpointManager().setRecoveryPath("managed_state_" + stream1State);
+    stream1Store.getCheckpointManager().setRecoveryPath("managed_state_" + stream2State);
+    stream1Store.setup(context);
+    stream2Store.setup(context);
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    stream1Store.beginWindow(windowId);
+    stream2Store.beginWindow(windowId);
+    super.beginWindow(windowId);
+  }
+
+  /**
+   * Process the waiting events
+   * @param finalize finalize Whether or not to wait for future to return
+   */
+  private void processWaitEvents(boolean finalize)
+  {
+    Iterator<Map.Entry<JoinEvent<K,T>, Future<List>>> waitIterator = waitingEvents.entrySet().iterator();
+    while (waitIterator.hasNext()) {
+      Map.Entry<JoinEvent<K,T>, Future<List>> waitingEvent = waitIterator.next();
+      Future<List> future = waitingEvent.getValue();
+      if (future.isDone() || finalize) {
+        try {
+          JoinEvent<K,T> event = waitingEvent.getKey();
+          joinStream(event.value,event.isStream1Data,future.get());
+        } catch (InterruptedException | ExecutionException e) {
+          throw new RuntimeException("end window", e);
+        }
+        waitIterator.remove();
+        if (!finalize) {
+          break;
```

Comment:
You can break only when you find a future which is not done... Till then you can keep processing for all the waitingEvents which has done future.